### ISSUE TITLE
Beat Saber Cannot be Modded due to WindHawk interfering with IPA.exe

### ISSUE THREAD

**Issue by IAMDEDJUSTTOGETFILES** [ramensoftware/windhawk#358]
Hello,
BSMG Support here. There's apparently an issue with WindHawk and BSIPA, our modloader for the pc version of the game, in which IPA.exe apparently tries to modify the necessary files, but when it tries to modify winhttp.dll, the permission is denied as seen in the attached image due to WindHawk. 

The issue can be fixed by putting IPA.exe in the process exclusions list or by deleting/removing WindHawk. Has this issue come up before? I'll try to get log files or anything you need to debug this.
![image (1)](https://github.com/user-attachments/assets/056bbbc4-fd1a-460e-9117-12e04e63bc69)

**Comment by m417z** [comment]
Hi, are you the author of IPA.exe? Can you reduce the issue to a small package that can be used to reproduce the problem? For example, attach the original or modified IPA.exe along with winhttp.dll and tell me how to run it to see the problem.

**Comment by Meivyn** [comment]
Hello, I am the maintainer of BSIPA.

We've had people complaining recently that the mod loader couldn't be installed, and this error was showing up. I asked them to send a minidump and I noticed that `winhttp.dll` was loaded by `IPA.exe` because of this binary being injected into the program:
![image](https://github.com/user-attachments/assets/09cc9515-80fa-442a-9009-3ffd42f0d601)

Here's the minidump file: https://pixeldrain.com/u/canuNJQY

Somehow the first report we've got about this issue happened *way* after the latest release you made, only about a month ago. And I didn't make any BSIPA release either in that timeframe, so I assume that it has something to do with system updates or Windhawk mods updates. I can't guarantee that it wasn't an issue before, but we didn't get a single report of it, and now we suddenly get a lot of them.

BSIPA uses an old customized version of the [Unity doorstop](https://github.com/NeighTools/UnityDoorstop) to inject the game and install mods. `IPA.exe` is just moving files and taking care of backups, it does nothing special: https://github.com/nike4613/BeatSaber-IPA-Reloaded/blob/master/IPA/Program.cs

You can get the latest release of BSIPA [here](https://github.com/nike4613/BeatSaber-IPA-Reloaded/releases/latest). It'll only work with a game made with Unity. Extract it into the game folder and run `IPA.exe` if you want to try it out.

**Comment by m417z** [comment]
> It'll only work with a game made with Unity. Extract it into the game folder and run `IPA.exe` if you want to try it out.

Can it be reproduced without having a game installed?

**Comment by Meivyn** [comment]
IPA won't work without a game, but I assume that any program trying to move this file would be enough to reproduce.

**Comment by m417z** [comment]
It'd be surprising if having winhttp.dll loaded affected moving a file of the same name. My best guess of how it can cause issues is that perhaps IPA loads its own winhttp.dll and does something like `GetModuleFileName("winhttp.dll")`, assuming only its copy is loaded. Then perhaps it's actually trying to delete the system winhttp.dll, which makes sense to result in an access denied error.

What I can do to mitigate the issue to have winhttp.dll delay-loaded. It's not used that often and it makes sense to do it to reduce the chance of conflicts, although if it does get loaded, the same issue will occur, so it's not a complete fix.

**Comment by m417z** [comment]
> perhaps IPA loads its own winhttp.dll

Indeed it does, as can be seen in the dump.

```
0:000> lmf
start             end                 module name
0000026a`ad3c0000 0000026a`ad3d4000   IPA      C:\Users\...\depots\620981\15813345\IPA.exe
[...]
00007ffa`c46d0000 00007ffa`c46dc000   winhttp  C:\Users\...\depots\620981\15813345\winhttp.dll
00007ffa`dbf60000 00007ffa`dbf6a000   version  C:\Windows\System32\version.dll
00007ffa`de530000 00007ffa`de666000   winhttp_7ffade530000 C:\Windows\System32\winhttp.dll
```

**Comment by Meivyn** [comment]
It does load it, but only because of the Windhawk DLL being injected into it. It doesn't otherwise. `winhttp.dll` is in the same directory as the program.

**Comment by m417z** [comment]
OK, now I understand the issue better. winhttp.dll isn't in the list of known DLLs, so it's loaded from the current directory, at which point it's in use so it can't be deleted.

Perhaps I can use delay loading together with [a custom handler](https://devblogs.microsoft.com/oldnewthing/20170126-00/?p=95265) to load the system32 version specifically. I'll see what I can do.

**Comment by Meivyn** [comment]
Not loading the DLL from the current directory would indeed solve the issue. However, I'm still flabbergasted that it didn't happen before considering that you are injecting all programs on purpose (at least from what I've understood of your [blog post](https://m417z.com/Implementing-Global-Injection-and-Hooking-in-Windows/)). I really wonder why.

**Comment by IAMDEDJUSTTOGETFILES** [comment]
We already know beat saber can be modded by putting IPA.exe in the exclusion list, as I said when I first created the post, we just want to know if it can be fixed or not.

**Comment by InterGalacticCold** [comment]
@IAMDEDJUSTTOGETFILES Have you tried putting IPA.exe in the exclusions list of windhawk?

**Comment by IAMDEDJUSTTOGETFILES** [comment]
Yes, as I have said before. I am going to say before you commit on issues, you should read the issue itself first. Mindlessly replying without reading leads you to miss Info that I have already said. Thank you.

**Comment by m417z** [comment]
Fixed in Windhawk v1.6.

**Comment by Meivyn** [comment]
Is this how that was fixed? Because this is definitely not an appropriate fix. Especially when user uses a downgraded version of the game that they can launch from anywhere, as long as the Steam environment variable is set. Plus that doesn't include other providers such as Oculus.

![Image](https://github.com/user-attachments/assets/f37cbcc6-c8d8-4453-8e5e-718d01cd8766)

**Comment by m417z** [comment]
@Meivyn no, that's unrelated. The `windhawk.dll` module that Windhawk injected into all processes had `winhttp.dll` in its import table, and because it's not in the list of known DLLs, it was loaded from the executable's directory or current directory, if a `winhttp.dll` file was found there.

Windhawk v1.6's `windhawk.dll` doesn't have `winhttp.dll` in its import table. Instead, it loads it on demand, using `LoadLibraryExW` with the `LOAD_LIBRARY_SEARCH_SYSTEM32` flag, which ensures that it's only loaded from the system32 folder.
